core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt

```diff
@@ -299,96 +303,86 @@ object UnitActions {
         }
     }
 
-    private fun addUnitUpgradeAction(unit: MapUnit, actionList: ArrayList<UnitAction>) {
-        val upgradeAction = getUpgradeAction(unit)
+    private fun addUnitUpgradeAction(
+        unit: MapUnit,
+        actionList: ArrayList<UnitAction>,
+        maxSteps: Int = Int.MAX_VALUE
+    ) {
+        val upgradeAction = getUpgradeAction(unit, maxSteps)
         if (upgradeAction != null) actionList += upgradeAction
     }
 
-    fun getUpgradeAction(unit: MapUnit): UnitAction? {
-        val tile = unit.currentTile
+    /**  Common implementation for [getUpgradeAction], [getFreeUpgradeAction] and [getAncientRuinsUpgradeAction] */
+    private fun getUpgradeAction(
+        unit: MapUnit,
+        maxSteps: Int,
+        isFree: Boolean,
+        isSpecial: Boolean
+    ): UnitAction? {
         if (unit.baseUnit().upgradesTo == null) return null
-        if (!unit.canUpgrade()) return null
-        if (tile.getOwner() != unit.civInfo) return null
-        
-        val upgradedUnit = unit.getUnitToUpgradeTo()
-        val goldCostOfUpgrade = unit.getCostOfUpgrade()
+        val unitTile = unit.getTile()
+        val civInfo = unit.civInfo
+        if (!isFree && unitTile.getOwner() != civInfo) return null
+
+        val upgradesTo = unit.baseUnit().upgradesTo
+        val specialUpgradesTo = unit.baseUnit().specialUpgradesTo
+        val upgradedUnit = when {
+            isSpecial && specialUpgradesTo != null -> civInfo.getEquivalentUnit (specialUpgradesTo)
+            isFree && upgradesTo != null -> civInfo.getEquivalentUnit(upgradesTo)  // getUnitToUpgradeTo can't ignore tech
+            else -> unit.getUnitToUpgradeTo(maxSteps)
+        }
+        if (!unit.canUpgrade(unitToUpgradeTo = upgradedUnit, ignoreRequirements = isFree, ignoreResources = true))
+            return null
 
-        return UnitAction(UnitActionType.Upgrade,
-            title = "Upgrade to [${upgradedUnit.name}] ([$goldCostOfUpgrade] gold)",
-            action = {
-                val unitTile = unit.getTile()
-                unit.destroy()
-                val newUnit = unit.civInfo.placeUnitNearTile(unitTile.position, upgradedUnit.name)
+        // Check _new_ resource requirements (display only - yes even for free or special upgrades)
+        // Using Counter to aggregate is a bit exaggerated, but - respect the mad modder.
+        val delta = Counter<String>()
+        for ((resource, amount) in upgradedUnit.getResourceRequirements())
+            delta.add(resource, amount)
+        val deltaStr = if (delta.all { it.value == 0 }) ""
```

Comment:
I think this is unnecessary since emptyList.joinToString will be empty, but I'm not sure

Reply:
"not sure"? You caught a waaaaaay  M A J O R blunder, thanks! :woman_facepalming: :see_no_evil: ... :id: :information_source: :zero: :t-rex:

Reply:
Actually I think the redundant all-zero test was an intended small optimization, maybe it had a tiny merit before I chose to use Counter. But the blunder is still right there in those four lines.